Support the use of the env-var KUBECONFIG

kenfdev opened the pull request:
This change respects the usage of the environment variable KUBECONFIG
when loading the kubeconfig. It also supports the default behavior of
setting multiple kubeconfig paths inside the KUBECONFIG variable.

Signed-off-by: Ken Fukuyama <kenfdev@gmail.com>

Related issue: https://github.com/solo-io/glooctl/issues/53

---

**This change hasn't been tested.**

I wanted to test what I have changed and also test this with `glooctl` but had no luck so far in building on my Mac. I wonder if this can be tested on a Mac in the first place (as documented, maybe this only works on Linux?).

I've tried following https://gloo.solo.io/dev/README/ but this didn't work either.

I'm more than happy to be able to contribute, but am stuck for now.

kenfdev commented:
This is what I get with `make localgloo` for now:

```
➜  gloo git:(multiple-kubeconfigs) ✗ make localgloo
go generate ./pkg/... ./internal/...
pkg/plugins/common/transformation/base_plugin.go:24: running "protoc": fork/exec /usr/local/bin/protoc: exec format error
pkg/plugins/connect/plugin.go:34: running "protoc": fork/exec /usr/local/bin/protoc: exec format error
pkg/plugins/nats-streaming/plugin.go:18: running "protoc": fork/exec /usr/local/bin/protoc: exec format error
"Sun, 01 Jul 2018 09:14:59 JST: github.com/solo-io/gloo/pkg/storage/generate/generate_clients.go:56"	success
"Sun, 01 Jul 2018 09:14:59 JST: github.com/solo-io/gloo/pkg/storage/generate/generate_clients.go:56"	success
"Sun, 01 Jul 2018 09:14:59 JST: github.com/solo-io/gloo/pkg/storage/generate/generate_clients.go:56"	success
2018/07/01 09:15:00 building a plan for generation
2018/07/01 09:15:00 planning definitions
2018/07/01 09:15:00 planning operations
2018/07/01 09:15:01 grouping operations into packages
2018/07/01 09:15:01 planning meta data and facades
2018/07/01 09:15:01 rendering 1 templates for model App
2018/07/01 09:15:01 name field App
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "app.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model AppWrapper
2018/07/01 09:15:01 name field AppWrapper
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "app_wrapper.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model AppsWrapper
2018/07/01 09:15:01 name field AppsWrapper
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "apps_wrapper.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model Error
2018/07/01 09:15:01 name field Error
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "error.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model ErrorBody
2018/07/01 09:15:01 name field ErrorBody
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "error_body.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model Route
2018/07/01 09:15:01 name field Route
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "route.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model RouteWrapper
2018/07/01 09:15:01 name field RouteWrapper
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "route_wrapper.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 1 templates for model RoutesWrapper
2018/07/01 09:15:01 name field RoutesWrapper
2018/07/01 09:15:01 package field models
2018/07/01 09:15:01 creating generated file "routes_wrapper.go" in "models" as definition
2018/07/01 09:15:01 executed template asset:model
2018/07/01 09:15:01 rendering 2 templates for operation
2018/07/01 09:15:01 name field GetApps
2018/07/01 09:15:01 package field apps
2018/07/01 09:15:01 creating generated file "get_apps_parameters.go" in "client/apps" as parameters
2018/07/01 09:15:01 executed template asset:clientParameter
2018/07/01 09:15:01 name field GetApps
2018/07/01 09:15:01 package field apps
2018/07/01 09:15:01 creating generated file "get_apps_responses.go" in "client/apps" as responses
2018/07/01 09:15:01 executed template asset:clientResponse
2018/07/01 09:15:01 rendering 1 templates for operation group
2018/07/01 09:15:01 name field apps
2018/07/01 09:15:01 creating generated file "apps_client.go" in "client/apps" as client
2018/07/01 09:15:01 executed template asset:clientClient
2018/07/01 09:15:01 rendering 2 templates for operation
2018/07/01 09:15:01 name field GetAppsAppRoutes
2018/07/01 09:15:01 package field routes
2018/07/01 09:15:01 creating generated file "get_apps_app_routes_parameters.go" in "client/routes" as parameters
2018/07/01 09:15:01 executed template asset:clientParameter
2018/07/01 09:15:01 name field GetAppsAppRoutes
2018/07/01 09:15:01 package field routes
2018/07/01 09:15:01 creating generated file "get_apps_app_routes_responses.go" in "client/routes" as responses
2018/07/01 09:15:01 executed template asset:clientResponse
2018/07/01 09:15:01 rendering 1 templates for operation group
2018/07/01 09:15:01 name field routes
2018/07/01 09:15:01 creating generated file "routes_client.go" in "client/routes" as client
2018/07/01 09:15:01 executed template asset:clientClient
2018/07/01 09:15:01 rendering 1 templates for application Fn
2018/07/01 09:15:01 name field Fn
2018/07/01 09:15:01 package field client
2018/07/01 09:15:01 creating generated file "fn_client.go" in "client" as facade
2018/07/01 09:15:01 executed template asset:clientFacade
2018/07/01 09:15:01 Generation completed!

For this generation to compile you need to have some packages in your GOPATH:

	* github.com/go-openapi/runtime
	* golang.org/x/net/context
	* golang.org/x/net/context/ctxhttp

You can get these now with: go get -u -f ./...
make: *** [_output/.generated-code] Error 1
```

kenfdev commented:
My bad...
I probably uninstalled the mac version and just followed the docs... should've realized.

The error continued but I did the following:

* installed the submodules (I just `git clone`d the project so forgot about this
* `git clone`d the `lyft` repository

And I think I succeeded building `localgloo`. I should be able to move forward 👍

kenfdev commented:
I've ran minikube and tried checking `RUN_KUBE_TESTS=1 ginkgo -r pkg/ internal/` to see the tests pass but I guess I need a valid `envoy` running.

```
[Fail] Xds RunXDS Server with invalid node name [BeforeEach] successfully bootstraps the envoy proxy
/Users/fukuyamaken/dev/src/github.com/solo-io/gloo/internal/control-plane/xds/xds_test.go:46

[Fail] Xds RunXDS Server with invalid node name [It] has a 500 route with body 'Invalid Envoy Bootstrap Configuration. Please refer to Gloo documentation https://gloo.solo.io/'
/Users/fukuyamaken/dev/src/github.com/solo-io/gloo/internal/control-plane/xds/xds_test.go:61
```

IIUC, there's only a linux version of envoy https://github.com/solo-io/gloo/releases

I'll try preparing a linux environment for this. This is going to take time...

If I take off the 'WIP' of the PR, are the tests going to run on CI? I guess I have to rebase and resolve the conflicts though.

ilackarms commented:
You can try running your tests with the flag `-skipPackage` or something like that. I'm on mobile right now but I know ginkgo has a flag for this. Most important are the test/kube_e2e tests. Those should run fine

kenfdev commented:
I've confirmed `RUN_KUBE_TESTS=1 ginkgo -r test/kube_e2e` pass with my laptop.

```
Ran 20 of 23 Specs in 210.915 seconds
SUCCESS! -- 20 Passed | 0 Failed | 0 Pending | 3 Skipped PASS
```

But oddly, sometimes the test fail because of the 5 minute timeout. It looks like the curl request doesn't receive 200 sometimes.

I want to add some more test cases specific to this change, but it's quite difficult because the e2e tests randomly fail on my laptop... Do you have any suggestions @ilackarms ? Thanks.

ilackarms commented:
@kenfdev our tests are a bit hacky. each test in the e2e covers a number of cases (with some redundancy between tests) by simply timing out when things dont work. this can be anything ranging from kube-dns issues, to bad config passed to envoy (an actual bug), etc.

the way we usually debug is by pausing tests when they get into that retry loop with a **Ctrl+z** and then use `kubectl logs <control plane, ingress>` and `kubectl exec -ti testrunner /bin/bash` to debug things. 

the CI build failed because the setup for our secrets (kubeconfig, etc) only works for branches, not pull requests (for security reasons). we should modify our CI to run the subset of tests that don't require secrets to run when it's a pull request.

for now i'll trigger a CI build that includes our secrets.

ilackarms commented on pkg/storage/crd/crd_storage_client.go:
```diff
@@ -23,6 +26,34 @@ type Client struct {
+	configOverrides := &clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: masterURL}}
+
+	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides).ClientConfig()
+
```
nit: remove newline

kenfdev commented on pkg/storage/crd/crd_storage_client.go:
```diff
@@ -23,6 +26,34 @@ type Client struct {
 }
 
+// GetConfig gets the kubernetes client config
+func GetConfig(masterURL, kubeconfigPath string) (*rest.Config, error) {
```
Yes, I agree. Thanks for pointing this out. I forgot to ask where to place it.

How about something like `pkg/utils/kube/config.go` ?

kenfdev commented:
Thank you for sharing how to do what! I'll give it a try next time.
Since it's late here, I'll fix the 2 problems tomorrow.

kenfdev commented:
I've changed the review points and added some tests.
I didn't add test cases for `internalCluster` and default `$HOME/.kube/config` because they depend on the environment the test runs on. 
Only tests mainly related to the `KUBECONFIG` environment variable ones are added. I've checked that the default `$HOME/.kube/config` also works in my environment (just didn't write tests for it).

kenfdev commented:
I added a commit for this change, but I will squash it if this change is confirmed 👍

ilackarms commented:
the changes look good. once we see this pass CI we can go ahead and merge. removing the WIP tag from this PR

ilackarms commented:
thanks @kenfdev !

kenfdev commented:
Thanks for your support @ilackarms , too!!
